byterun/win32.c

```diff
@@ -57,10 +57,16 @@
 #endif
 
 /* Very old Microsoft headers don't include intptr_t */
-#if defined(_MSC_VER) && !defined(_UINTPTR_T_DEFINED)
-typedef unsigned int uintptr_t;
+#ifdef _MSC_VER
+#ifndef _UINTPTR_T_DEFINED
+typedef uintnat uintptr_t;
 #define _UINTPTR_T_DEFINED
 #endif
+#ifndef _INTPTR_T_DEFINED
+typedef intnat intptr_t;
```

Comment:
We still have the following in channels.c:
````
#if defined(_MSC_VER) && !defined(_INTPTR_T_DEFINED)
typedef int intptr_t;
#define _INTPTR_T_DEFINED
#endif
....
...
int win_CRT_fd_of_filedescr(value handle)
{
  if (CRT_fd_val(handle) != NO_CRT_FD) {
    return CRT_fd_val(handle);
  } else {
    int fd = _open_osfhandle((intptr_t) Handle_val(handle), O_BINARY);
    if (fd == -1) uerror("channel_of_descr", Nothing);
    CRT_fd_val(handle) = fd;
    return fd;
  }
}
````

Wouldn't that cause a similar problem, on 64-bit and old enough versions of MSVC?  (Or perhaps all 64-bit versions do define intptr_t?)

Reply:
The `typedef` is wrong, but in principle it is safe: file `HANDLE`s in Win 64 are still 32-bits. See https://msdn.microsoft.com/en-us/library/windows/desktop/aa384203%28v=vs.85%29.aspx.

Reply:
Ok.  Perhaps worth fixing it, but can be done independently of this PR (unless we prefer to add the missing typedef in a central place).

Reply:
What about adding after the definition intnat and uintnat in config.h the following piece of code:
````
#ifndef _UINTPTR_T_DEFINED
typedef uintnat uintptr_t;
#define _UINTPTR_T_DEFINED
#endif
#ifndef _INTPTR_T_DEFINED
typedef intnat intptr_t;
#define _INTPTR_T_DEFINED
#endif
```
and remove all other typedef of intptr_t in the codebase ?

Reply:
Or could simply use "intnat" and "uintnat" in place of intptr_t ?
(The only drawbacks I can think of is the risk to forget to replace it in future code).

Reply:
> Or could simply use "intnat" and "uintnat" in place of intptr_t ?

That might be more prudent than defining `{u,}intptr_t` yourself, with the risk that some system header file doesn't understand what you did and tries to define it again.

On the other hand, if all recent enough versions of MSVC define those standard types, maybe we should just give up on supporting  the older versions.

Reply:
I full agree.  So, we need to decide between:

  - Use uintnat/intnat (unconditionally)
  - Use intptr_t/uintptr_t, which means no longer supporting old versions of MSVC

My understanding is that versions of MSVC that don't support intptr_t are also versions that don't come with a 64-bit variant, meaning they are really old.  Support could be restored purely in the build system by passing something like `/Dintptr_t=int` to cl.exe, if really needed.

@dra27 : you would be the (likely only) one to care for such old versions, so your input would be much appreciated!